lib/accesslists/hierarchy.go

```diff
@@ -42,28 +42,100 @@ const (
 type AccessListAndMembersGetter interface {
 	ListAccessListMembers(ctx context.Context, accessListName string, pageSize int, pageToken string) (members []*accesslist.AccessListMember, nextToken string, err error)
 	GetAccessList(ctx context.Context, accessListName string) (*accesslist.AccessList, error)
 	GetAccessListMember(ctx context.Context, accessList string, memberName string) (*accesslist.AccessListMember, error)
 }
 
+type ancestorOptions struct {
+	validateUserRequirement bool
+	clock                   clockwork.Clock
+	user                    types.User
+}
+
+func (o *ancestorOptions) validate() error {
+	if o.validateUserRequirement {
+		if o.user == nil {
+			return trace.BadParameter("user is required when validateUserRequirement is true")
+		}
+		if o.clock == nil {
+			o.clock = clockwork.NewRealClock()
+		}
+	}
+	return nil
+}
+
+// ancestorOption is a functional option for configuring the behavior of GetAncestorsFor.
+type ancestorOption func(*ancestorOptions)
+
+func withUserRequirementsCheck(user types.User, clock clockwork.Clock) ancestorOption {
+	return func(opts *ancestorOptions) {
+		opts.validateUserRequirement = true
+		opts.user = user
+		opts.clock = clock
+	}
+}
+
+// HierarchyConfig holds dependencies for building access list hierarchies.
+type HierarchyConfig struct {
+	// AccessListService is used to fetch Access Lists and their members.
+	AccessListsService AccessListAndMembersGetter
+	// Getter is used to fetch Access Lists and their members.
+	Clock clockwork.Clock
+	// LockService is used to fetch user locks.
+	LockService services.LockGetter
+}
+
+// CheckAndSetDefaults validates the config and sets default values.
+func (c *HierarchyConfig) CheckAndSetDefaults() error {
+	if c.AccessListsService == nil {
+		return trace.BadParameter("AccessListsService is required")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// Hierarchy provides methods to compute access list hierarchies.
+type Hierarchy struct {
+	HierarchyConfig
+}
+
+// NewHierarchy constructs a HierarchyService with the given config.
+func NewHierarchy(cfg HierarchyConfig) (*Hierarchy, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Hierarchy{
+		HierarchyConfig: cfg,
+	}, nil
+}
+
 // GetMembersFor returns a flattened list of Members for an Access List, including inherited Members.
 //
 // Returned Members are not validated for expiration or other requirements – use IsAccessListMember
 // to validate a Member's membership status.
-func GetMembersFor(ctx context.Context, accessListName string, g AccessListAndMembersGetter) ([]*accesslist.AccessListMember, error) {
-	return getMembersFor(ctx, accessListName, g, make(map[string]struct{}))
+func (s *Hierarchy) GetMembersFor(ctx context.Context, accessListName string) ([]*accesslist.AccessListMember, error) {
```

Comment:
```suggestion
func (h *Hierarchy) GetMembersFor(ctx context.Context, accessListName string) ([]*accesslist.AccessListMember, error) {
```

Applies to all other recieivers